packages/admin/cms-admin/src/pages/pageTree/useExtractPages.tsx

```diff
@@ -0,0 +1,275 @@
+
+const updatePageNodeMutation = gql`
+    mutation ImportContentsUpdatePageNode($nodeId: ID!, $input: PageTreeNodeUpdateInput!) {
+        updatePageTreeNode(id: $nodeId, input: $input) {
+            id
+            name
+            slug
+        }
+    }
+`;
+
+const transformToSlug = (name: string, locale: string) => {
+    let slug = slugify(name, { replacement: "-", lower: true, locale });
+    // Remove everything except unreserved characters and percent encoding (https://tools.ietf.org/html/rfc3986#section-2.1)
+    // This is necessary because slugify does not remove all reserved characters per default (e.g. "@")
+    slug = slug.replace(/[^a-zA-Z0-9-._~]/g, "");
+    return slug;
+};
+
+interface UseExtractPagesApi {
+    /**
+     * Iterates over passed pages synchronous and extracts contents
+     *
+     * Process:
+     *      1. traverses the tree with top-down strategy
+     *          1a. documents and blocks return their translatable text contents
+     *      2. returns stringified JSON including all extracted text contents
+     *
+     * @param flatPagesTree
+     */
+    extractContents: (pages: GQLPageTreePageFragment[]) => Promise<string[]>;
+
+    /**
+     * Iterates over passed pages synchronous and updates data with mutations
+     *
+     * Process:
+     *      1. traverses the tree with top-down strategy
+     *          1a. documents and blocks replace old content in their state with new content
+     *          1b. updates Page
+     *          1c. updates Page Node
+     *
+     * @param flatPagesTree
+     * @param contents parsed contents data from clipboard
+     */
+    importContents: (pages: GQLPageTreePageFragment[], contents: ParsedContents) => Promise<void>;
+
+    /**
+     * read data from clipboard, validate it and return parsed data.
+     */
+    getContentsFromClipboard: () => Promise<GetContentsFromClipboardResponse>;
+}
+
+/**
+ * This hooks provides some helper functions to extract Pages and PageTreeNodes
+ */
+function useExtractImportPages(): UseExtractPagesApi {
+    const { documentTypes } = usePageTreeContext();
+    const client = useApolloClient();
+    const { scope } = useContentScope();
+    const locale = useLocale({ scope });
+
+    const extractContents = React.useCallback(
+        async (pages: GQLPageTreePageFragment[]): Promise<string[]> => {
+            const textContents: string[] = [];
+
+            await Promise.all(
+                pages.map(async (page) => {
+                    const documentType = documentTypes[page.documentType];
+
+                    if (!documentType) {
+                        throw new Error(`Unknown document type "${documentType}"`);
+                    }
+
+                    try {
+                        const query = documentType.getQuery;
+
+                        if (query) {
+                            const { data } = await client.query<GQLPageQuery, GQLPageQueryVariables>({
```

Comment:
`GQLPageQuery` is misleading here, as that can be any document type. I'd say any/unknown is appropriate here.